kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java

```diff
@@ -52,6 +63,150 @@ public void testForwardRequest() {
+        Thread.sleep(60L);
+        channel.runPendingTasks();
+
+        assertThat(filterFuture).isCompletedExceptionally().isNotCancelled();
+        assertThatThrownBy(filterFuture::get).hasCauseInstanceOf(TimeoutException.class);
+        assertThat(channel.isOpen()).isFalse();
+    }
+
+    @Test
+    public void testDeferredResponseTimeout() throws Exception {
+        var filterFuture = new CompletableFuture<ResponseFilterResult>();
+        ApiVersionsResponseFilter filter = (apiVersion, header, request, context) -> filterFuture;
+        buildChannel(filter, 50L);
+        writeResponse(new ApiVersionsResponseData());
+        Thread.sleep(60L);
+        channel.runPendingTasks();
+
+        assertThat(filterFuture).isCompletedExceptionally().isNotCancelled();
+        assertThatThrownBy(filterFuture::get).hasCauseInstanceOf(TimeoutException.class);
+        assertThat(channel.isOpen()).isFalse();
+    }
+
+    static Stream<Arguments> requestFilterClosesChannel() {
+        return Stream.of(
+                Arguments.of("completes exceptionally",
+                        (BiFunction<RequestHeaderData, ApiMessage, CompletionStage<RequestFilterResult>>) (header, request) -> CompletableFuture
+                                .failedStage(new RuntimeException("filter error")),
+                        false),
+                Arguments.of("filter result signals close",
+                        (BiFunction<RequestHeaderData, ApiMessage, CompletionStage<RequestFilterResult>>) (header, request) -> new RequestFilterResultBuilderImpl()
+                                .withCloseConnection().completed(),
+                        false),
+                Arguments.of("filter result signals close with forward",
+                        (BiFunction<RequestHeaderData, ApiMessage, CompletionStage<RequestFilterResult>>) (header, request) -> new RequestFilterResultBuilderImpl()
+                                .forward(header, request).withCloseConnection().completed(),
+                        true));
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource
+    void requestFilterClosesChannel(String name,
+                                    BiFunction<RequestHeaderData, ApiMessage, CompletableFuture<RequestFilterResult>> stageFunction,
+                                    boolean forwardExpected) {
+        ApiVersionsRequestFilter filter = (apiVersion, header, request, context) -> stageFunction.apply(header, request);
+        buildChannel(filter);
+        var frame = writeRequest(new ApiVersionsRequestData());
+        channel.runPendingTasks();
+
+        assertThat(channel.isOpen()).isFalse();
+        var propagated = channel.readOutbound();
+        if (forwardExpected) {
+            assertThat(propagated).isEqualTo(frame);
+        }
+        else {
+            assertThat(propagated).isNull();
+        }
+    }
+
+    static Stream<Arguments> responseFilterClosesChannel() {
+        return Stream.of(
```

Comment:
Should we be covering `ShortCircuit` and `Drop` here as well?

Reply:
drop doesn't combine with close (I assume because drop+close has the same effect as close).

but yes I think we could have a test for a short-circuit response + close (though it's a request test, so the other `Stream<Arguments>` method)

Reply:
Both good points...

Reply:
I added a unit test to my PR `#requestShortCircuit`